File: copasi/bindings/cpp_examples/example2/example2.cpp
// Copyright (C) 2019 - 2021 by Pedro Mendes, Rector and Visitors of the
// University of Virginia, University of Heidelberg, and University
// of Connecticut School of Medicine.
// All rights reserved.

// Copyright (C) 2017 - 2018 by Pedro Mendes, Virginia Tech Intellectual
// Properties, Inc., University of Heidelberg, and University of
// of Connecticut School of Medicine.
// All rights reserved.

// Copyright (C) 2010 - 2016 by Pedro Mendes, Virginia Tech Intellectual
// Properties, Inc., University of Heidelberg, and The University
// of Manchester.
// All rights reserved.

// Copyright (C) 2009 by Pedro Mendes, Virginia Tech Intellectual
// Properties, Inc., EML Research, gGmbH, University of Heidelberg,
// and The University of Manchester.
// All rights reserved.

/**
 * This is an example on how to load a cps file
 * and output some information on the model
 */

#include <iostream>
#include <string>

#define COPASI_MAIN

#include "copasi/copasi.h"
#include "copasi/core/CRootContainer.h"
#include "copasi/CopasiDataModel/CDataModel.h"
#include "copasi/model/CModel.h"
#include "copasi/model/CCompartment.h"
#include "copasi/model/CMetab.h"
#include "copasi/model/CReaction.h"

int main(int argc, char** argv)
{
  // initialize the backend library
  CRootContainer::init(argc, argv);
  assert(CRootContainer::getRoot() != NULL);
  // create a new datamodel
  CDataModel* pDataModel = CRootContainer::addDatamodel();
  assert(CRootContainer::getDatamodelList()->size() == 1);

  // the only argument to the main routine should be the name of a CPS file
  if (argc == 2)
    {
      std::string filename = argv[1];

      try
        {
          // load the model without progress report
          pDataModel->loadFromFile(filename);
        }
      catch (...)
        {
          std::cerr << "Error while loading the model from file named \"" << filename << "\"." << std::endl;
          CRootContainer::destroy();
          return 1;
        }

      CModel* pModel = pDataModel->getModel();
      assert(pModel != NULL);
      std::cout << "Model statistics for model \"" << pModel->getObjectName() << "\"." << std::endl;

      // output number and names of all compartments
      size_t i, iMax = pModel->getCompartments().size();
      std::cout << "Number of Compartments: " << iMax << std::endl;
      std::cout << "Compartments: " << std::endl;

      for (i = 0; i < iMax; ++i)
        {
          CCompartment* pCompartment = &pModel->getCompartments()[i];
          assert(pCompartment != NULL);
          std::cout << "\t" << pCompartment->getObjectName() << std::endl;
        }

      // output number and names of all metabolites
      iMax = pModel->getMetabolites().size();
      std::cout << "Number of Metabolites: " << iMax << std::endl;
      std::cout << "Metabolites: " << std::endl;

      for (i = 0; i < iMax; ++i)
        {
          CMetab* pMetab = &pModel->getMetabolites()[i];
          assert(pMetab != NULL);
          std::cout << "\t" << pMetab->getObjectName() << std::endl;
        }

      // output number and names of all reactions
      iMax = pModel->getReactions().size();
      std::cout << "Number of Reactions: " << iMax << std::endl;
      std::cout << "Reactions: " << std::endl;

      for (i = 0; i < iMax; ++i)
        {
          CReaction* pReaction = &pModel->getReactions()[i];
          assert(pReaction != NULL);
          std::cout << "\t" << pReaction->getObjectName() << std::endl;
        }
    }
  else
    {
      std::cerr << "Usage: example2 CPSFILE" << std::endl;
      CRootContainer::destroy();
      return 1;
    }

  // clean up the library
  CRootContainer::destroy();
}

File: copasi/bindings/R/examples/example2.R
#options(echo = FALSE) # disable echoing of input

#
# This is an example on how to load a cps file
# and output some information on the model
#

# First we load the COPASI package
dyn.load(paste("COPASI", .Platform$dynlib.ext, sep=""))
source("COPASI.R")
# I Don't know exactly what the next line does, but this is what the SWIG
# documentation has to say about it:
# The cacheMetaData(1) will cause R to refresh its object tables. Without it, inheritance of wrapped objects may fail.
cacheMetaData(1)

stopifnot(!is.null(CRootContainer_getRoot()))
# create a datamodel
dataModel <- CRootContainer_addDatamodel()
stopifnot(CRootContainer_getDatamodelList()$size() == 1)
# the only argument to the main routine should be the name of a CPS file
args <- commandArgs(trailingOnly = TRUE)
if (length(args) == 1) {
    filename <- args[1]
    # load the model without progress report
   
    # I have no clue how exception handling in R works
    tryCatch(dataModel$loadModel(filename), error = function(e) {
      write(paste("Error while loading the model from file named \"" , filename , "\"."), stderr())
      quit(save = "default", status = 1, runLast = TRUE)
    } )

    model <- dataModel$getModel()
    stopifnot(!is.null(model))
    cat('Model statistics for model "' , model$getObjectName() , '".\n', sep="")

    # output number and names of all compartments
    iMax <- model$getCompartments()$size()
    cat("Number of Compartments: " , iMax, "\n", sep="")
    cat("Compartments: \n")
    i <- 0
    while ( i < iMax) {
        compartment <- model$getCompartment(i)
        stopifnot(!is.null(compartment))
        cat("    " , compartment$getObjectName(), "\n", sep = "")
        i <- i + 1
    }
    # output number and names of all metabolites
    iMax <- model$getMetabolites()$size()
    cat("Number of Metabolites: " , iMax, "\n", sep = "")
    cat("Metabolites: \n")
    i <- 0
    while (i < iMax) {
        metab <- model$getMetabolite(i)
        stopifnot(!is.null(metab))
        cat("    " , metab$getObjectName(), "\n", sep="")
        i <- i + 1
    }
    # output number and names of all reactions
    iMax <- model$getReactions()$size()
    cat("Number of Reactions: " , iMax, "\n", sep = "")
    cat("Reactions: \n")
    i <- 0
    while ( i < iMax) {
        reaction <- model$getReaction(i)
        stopifnot(!is.null(reaction))
        cat("    " , reaction$getObjectName() , "\n", sep="")
        i <- i + 1
    }
} else {
    write( "Usage: example2 CPSFILE", stderr())
    quit(save = "default", status = 1, runLast = TRUE)
}


